constantine-go/constantine.go

```diff
@@ -265,4 +266,428 @@ func (ctx EthKzgContext) VerifyBlobKzgProofBatchParallel(tp Threadpool, blobs []
+		(*C.ctt_span)(unsafe.Pointer(&spans[0])),
+		(*C.ctt_eth_bls_signature)(unsafe.Pointer(&signatures[0])),
+		(C.ptrdiff_t)(len(pubkeys)),
+		(*C.byte)(unsafe.Pointer(&secureRandomBytes[0])),
+	)
+	if status != C.cttEthBls_Success {
+		err := errors.New(
+			C.GoString(C.ctt_eth_bls_status_to_string(status)),
+		)
+		return false, err
+	}
+
+	return true, nil
+}
+
+func BatchVerifyParallel(tp Threadpool, pubkeys []EthBlsPubKey, messages [][]byte, signatures []EthBlsSignature, secureRandomBytes [32]byte) (bool, error) {
+	if len(pubkeys) != len(messages) {
+		err := errors.New("Number of public keys must match number of messages.")
+		return false, err
+	} else if len(pubkeys) != len(signatures) {
+		err := errors.New("Number of public keys must match number of signatures.")
+		return false, err
+	}
+
+	// copy messages to CttSpan type array
+	spans := newSpans(messages)
+	defer freeSpans(spans) // make sure to free after!
+
+	status := C.ctt_eth_bls_batch_verify_parallel(tp.ctx,
+		(*C.ctt_eth_bls_pubkey)(unsafe.Pointer(&pubkeys[0])),
+		(*C.ctt_span)(unsafe.Pointer(&spans[0])),
+		(*C.ctt_eth_bls_signature)(unsafe.Pointer(&signatures[0])),
+		(C.ptrdiff_t)(len(pubkeys)),
+		(*C.byte)(unsafe.Pointer(&secureRandomBytes[0])),
+	)
+	if status != C.cttEthBls_Success {
+		err := errors.New(
+			C.GoString(C.ctt_eth_bls_status_to_string(status)),
+		)
+		return false, err
+	}
+
+	return true, nil
+}
+
+// NOTE: C.ctt_eth_bls_batch_sig_accumulator is an incomplete struct. Therefore
+// we use 2 functions on the Nim side to (de)allocate storage for the struct.
+type EthBlsBatchSigAccumulator struct {
+	ctx *C.ctt_eth_bls_batch_sig_accumulator
+}
+func EthBlsBatchSigAccumulatorAlloc() (accum EthBlsBatchSigAccumulator) {
+	accum.ctx = C.ctt_eth_bls_alloc_batch_sig_accumulator()
+	return accum
+}
+
+func EthBlsBatchSigAccumulatorFree(accum EthBlsBatchSigAccumulator) {
+	C.ctt_eth_bls_free_batch_sig_accumulator(accum.ctx)
+}
+
+func (accum EthBlsBatchSigAccumulator) Init(secureRandomBytes [32]byte) {
+	C.ctt_eth_bls_init_batch_sig_accumulator((*C.ctt_eth_bls_batch_sig_accumulator)(accum.ctx),
+		(*C.byte)(unsafe.Pointer(&secureRandomBytes[0])),
+	)
+}
+
+func (accum EthBlsBatchSigAccumulator) Update(pub EthBlsPubKey, message []byte, sig EthBlsSignature) bool {
+	status := C.ctt_eth_bls_update_batch_sig_accumulator((*C.ctt_eth_bls_batch_sig_accumulator)(accum.ctx),
+		(*C.ctt_eth_bls_pubkey)(&pub),
+		(*C.byte)(unsafe.Pointer(&message[0])),
+		(C.ptrdiff_t)(len(message)),
+		(*C.ctt_eth_bls_signature)(&sig),
+	)
+	return bool(status)
+}
+
+func (accum EthBlsBatchSigAccumulator) FinalVerify() bool {
+	status := C.ctt_eth_bls_final_verify_batch_sig_accumulator(
+		(*C.ctt_eth_bls_batch_sig_accumulator)(accum.ctx),
+	)
+	return bool(status)
+}
+
+func BatchVerifyGo(pubkeys []EthBlsPubKey, messages [][]byte, signatures []EthBlsSignature, secureRandomBytes [32]byte) (bool, error) {
```

Comment:
And we can rename this to BatchVerifySoA, and we likely want a BatchVerifyAoS with an arrays of triplets (pubkey, message, signature).

And we can reimplement AggregateVerify in terms of the BLS Sig accumulator.

The parallel BatchVerify is not a priority and can be done later, probably using goroutines as it doesn't require advanced multithreading capabilities, see impl https://github.com/mratsim/constantine/blob/285aeb280f84b889195529bf821e2801423bcb08/constantine/signatures/bls_signatures_parallel.nim#L71-L181

The Go API is important for fuzzing in go-ethereum/geth but a parallel implementation is not needed for fuzzing.

Reply:
Renamed to `BatchVerifySoA` and added a `BatchVerifyAoS` with a `[]BatchVerifyTriplet` type as the argument.

Not quite sure how to reimplement `AggregateVerify` using the batch sig accumulator, because of the secure random bytes / my naive assumption of just reusing the aggregate signature for each (pkey/message) pair not actually giving the right result.

Reply:
For future reference: We'll expose the aggregate accumulator on the Nim side similar to the batch sig accumulator at a later date.